src/Strategy/CommonClassesStrategy.php

```diff
@@ -107,9 +112,20 @@ final class CommonClassesStrategy implements DiscoveryStrategy
      */
     public static function getCandidates($type)
     {
-        if (Psr18Client::class === $type) {
-            $candidates = self::$classes[PSR18Client::class];
+        $candidates = [];
 
+        foreach (self::$classes[$type] ?? [] as $c) {
+            // Guzzle 6 does not implement the PSR-18 client interface, but Guzzle 7 does.
+            if (Psr18Client::class === $type && GuzzleHttp::class === $c['class']) {
```

Comment:
i find this hard to read.

what if we would do

```
$candidates = self::$classes[$type] ?? [];
if (Psr18Client::class === $type) {
    if ( !defined(GuzzleHttp\ClientInterface::class.'::MAJOR_VERSION') && $pos = array_search(GuzzleHttp::class, $candidates)) {
        unset($candidates[$pos]);
    }
    foreach (self::$classes[HttpClient::class] as $c) {
        ...
    }
}
```

that way, the special case of psr18 with guzzle is more clear. we don't do the if in each iteration, and don't loop when we are not about psr18 client.

Reply:
or even `if (Psr18Client::class !== $type) return $c;` for an early return.

Reply:
We don't want an early return. I think what I wrote is the most readable out of everything there. I think it's pretty clear. We start with an empty array of candidates, and then we add in the ones we want, then return at the bottom of the function, in all cases.

Reply:
sorry, i want to continue challenging this bit of code a bit.

i find it difficult to see what exactly we do, and part of the `if` condition inside the loop is not depending on anything in the loop. for non psr18 client we do not need to loop at all.

how about this:

```
if (Psr18Client::class !== $type) {
    return self::$classes[$type] ?? [];;
}

$candidates = [];
foreach (self::$classes[Psr18Client::class] as $c) {
    if (GuzzleHttp::class !== $c || defined(GuzzleHttp\ClientInterface::class.'::MAJOR_VERSION')) {
        $candidates[] = $c;
    }
    // HTTPlug 2.0 clients implements PSR18Client too.
    foreach (self::$classes[HttpClient::class] as $c) {
        ...
    }
}
```

Reply:
I've updated it again, to handle PSR-18 stuff totally separately, to avoid confusion. Is that more to your taste now?

```php

    /**
     * {@inheritdoc}
     */
    public static function getCandidates($type)
    {
        if (Psr18Client::class === $type) {
            return self::getPsr18Candidates();
        }

        return self::$classes[$type] ?? [];
    }

    /**
     * @return array The return value is always an array with zero or more elements. Each
     *               element is an array with two keys ['class' => string, 'condition' => mixed].
     */
    private static function getPsr18Candidates()
    {
        $candidates = [];

        // Guzzle 6 does not implement the PSR-18 client interface, but Guzzle 7 does.
        foreach (self::$classes[Psr18Client::class] ?? [] as $c) {
            if (GuzzleHttp::class === $c['class']) {
                if (defined('GuzzleHttp\ClientInterface::MAJOR_VERSION')) {
                    $candidates[] = $c;
                }
            } else {
                $candidates[] = $c;
            }
        }

        // HTTPlug 2.0 clients implements PSR18Client too.
        foreach (self::$classes[HttpClient::class] as $c) {
            try {
                if (is_subclass_of($c['class'], Psr18Client::class)) {
                    $candidates[] = $c;
                }
            } catch (\Throwable $e) {
                trigger_error(sprintf('Got exception "%s (%s)" while checking if a PSR-18 Client is available', get_class($e), $e->getMessage()), E_USER_WARNING);
            }
        }

        return $candidates;
    }
```